### ISSUE TITLE
[HOLD for payment 2023-01-23] [$1000] `Add a payment method...` message disappears when the app is reloaded after deleting a bank offline

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14046]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Actions Performed:

**Web**

1. Add only one bank through Plaid
2. Turn the internet off
3. Go to Settings > Payments
4. Click on the bank
5. Click Delete to delete the bank
6. Close the tab
7. Turn the internet on
8. Open a new tab and run the app again
9. Go to Settings > Payments

**Mobile app**

1. Add only one bank through Plaid
2. Turn the internet off
3. Go to Settings > Payments
4. Tap on the bank
5. Tap Delete to delete the bank
6. Kill the app
7. Turn the internet on
8. Open the app again
9. Go to Settings > Payments

## Expected Result:
The add payment method message is displayed
## Actual Result:
The add payment method message is not displayed
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.2.49-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/13783842/210961067-c47ed1ba-4644-47ef-94b8-e4686fc42a72.mp4

https://user-images.githubusercontent.com/13783842/210961489-e252766d-f6ac-4d76-8c1f-ccb2add4300e.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @adeel0202 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1672901401567179

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014e025b9cf1e4c7a8</li>
        <li>Upwork Job ID: 1611272610208215040</li>
        <li>Last Price Increase: 2023-01-06</li>
    </ul>
</details>

**Comment by jasperhuangg** [comment]
This flow actually works if you don't reload the page, which is actually how we expect it to work.

https://user-images.githubusercontent.com/31285285/210922363-e35fc8cb-1af9-4da8-a695-21e5a12e7107.mov

When you reload the page, you queue an extra call to `OpenApp` before we call `ReconnectApp` that messes up the Onyx stores for any items that are pending deletion. Technically, when you're offline, you shouldn't be able to reload the page and queue that `OpenApp` call at all–this is just a quirk of how our "Force Offline" works that allows you to reload the page, but in production this situation should never happen.

Closing.

**Comment by jasperhuangg** [comment]
Actually, let me just confirm the differences between the two flows (reloading and not reloading) before I close this out for sure

**Comment by jasperhuangg** [comment]
So if you reload the page and then come back online, here's what happens:

1. Reloading the page queues an extra call to `OpenPaymentsPage` ([a read](https://github.com/Expensify/App/blob/main/src/libs/actions/App.js#L121-L122)), which updates the `bankAccountList` in Onyx before the offline `write` call `DeletePaymentBankAccount` can happen. Reads should ALWAYS happen after writes when coming back online.
2. Since the bank account hasn't been deleted in the back-end, nothing changes in Onyx. 
3. Then, you come back online, triggering the call to `DeletePaymentBankAccount`, and another call to `OpenPaymentsPage`.
4. Since a call to `OpenPaymentsPage` happened earlier, we're unable to clear the deleted bank account from the Onyx store, and we're left with a bank account with `pendingAction: 'delete'` in Onyx, which prevents us from showing the "Add a payment method..." text, which only appears if the `bankAccountList` is empty.

If we get rid of step 1, then this entire flow works. As for why this happens specifically, I'm still not entirely sure–I just know in production it's impossible to queue that first call to `OpenPaymentsPage` before we come back online, which means this scenario will never happen.

**Comment by jasperhuangg** [comment]
Reopening per https://expensify.slack.com/archives/C049HHMV9SM/p1672986568755309?thread_ts=1672901401.567179&cid=C049HHMV9SM

**Comment by hungvu193** [comment]
When you deleted a BankAccount in offline mode, the pendingAction in that BankAccount turn into 'delete'.
When you back online, data from server return the BankAccount without pendingAction property (See attachment below).
<img width="1680" alt="Screen Shot 2023-01-06 at 16 07 40" src="https://user-images.githubusercontent.com/16502320/210968543-9f1490c2-706e-40dd-9119-df5af5768e16.png">
Since it's a merge action, it combines the local bankAccountList and the state from server, so even you are back online, the pendingAction is still `delete`, because of that the OfflineWithFeedback will hide the children.
I think this bug can fix from the backend, we need to pass the pendingAction: undefined from server.

**Comment by bernhardoj** [comment]
### **Proposal**

I think it happens because 
I think the simplest way to solve this is to pass a success data to the `API.write`.

```diff
function deletePaymentBankAccount(bankAccountID) {
    API.write('DeletePaymentBankAccount', {
        bankAccountID,
    }, {
        optimisticData: [
            {
                onyxMethod: CONST.ONYX.METHOD.MERGE,
                key: `${ONYXKEYS.BANK_ACCOUNT_LIST}`,
                value: {[bankAccountID]: {pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE}},
            },
        ],
+       successData: [
+           {
+               onyxMethod: CONST.ONYX.METHOD.MERGE,
+               key: `${ONYXKEYS.BANK_ACCOUNT_LIST}`,
+               value: {[bankAccountID]: null},
+           },
+       ],
+   });
}
```

Without providing the success data, we rely on Pusher notification to update the onyx data when success. I think the problem with this is the persisted request is successfully executed before the pusher subscription complete, therefore we didn't receive any update. Maybe we can wait for the subscription to complete before hiding the splash screen, but we also need to do this only when the user is logged in. Another way is to wait the completion before executing any `SequentialQueue.push` (like what we do with offline condition). But again, there could be possibility where the subscription is failed, and the network request won't work at all. 

It is also possible that this case could happen on other part of the app feature.



**Result**

https://user-images.githubusercontent.com/50919443/211011743-d814c51e-cc51-435c-aaa0-d3cba84a71d2.mp4

**Comment by abdulrahuman5196** [comment]
**Proposal**

**Root Cause**
In PaymentMethodList.js we have set the add payment method message to be displayed as 'ListEmptyComponent' which will only display when we provide empty data to FlatList. But in this case when we delete the payment method and go offline, we still have the payment method data with "pendingAction": "delete" value as noted below which will only go away after we get response from server(also request still in network queue as well). So we still pass a data with payment entry to FlatList and flatList doesn't show the add payment menthod message.

```
{
  "title": "Plaid Checking",
  "description": "Account ending in 0000",
  "methodID": 1854845,
  "key": "bankAccount-1854845",
  "accountType": "bankAccount",
  "isDefault": false,
  **"pendingAction": "delete",**
  "iconSize": 40,
  "iconFill": null,
  "wrapperStyle": null
}
```
Note: leaving extra data as not required.

**Why is the single payment data entry with pendingAction not shown?**
We send a single payment method data to Flatlist which uses OfflineWithFeedback component for each row which hides its children when it has pendingAction of 'delete' - https://github.com/Expensify/App/blob/main/src/components/OfflineWithFeedback.js#L90. So visually the row doesn't have anything.


**Will the issue happen at different scenarios as well?**
Yes. you can even repro the issue without going offline. If you are on a slow network, you can notice a time delay after you delete the payment method and the add payment message shows up. This is also due to the same RCA above mentioned above and the UI becomes stale until we a get response from server.

**Solution**
We already filter payment methods based on different criteria before we sent it to FlatList here - https://github.com/Expensify/App/blob/main/src/pages/settings/Payments/PaymentMethodList.js#L116.
We can add an extra filter to ignore payment methods with pendingAction with the below code. 

```
--- a/src/pages/settings/Payments/PaymentMethodList.js
+++ b/src/pages/settings/Payments/PaymentMethodList.js
@@ -121,7 +121,7 @@ class PaymentMethodList extends Component {
         if (!_.isEmpty(this.props.filterType)) {
             combinedPaymentMethods = _.filter(combinedPaymentMethods, paymentMethod => paymentMethod.accountType === this.props.filterType);
         }
-
+        combinedPaymentMethods = _.filter(combinedPaymentMethods, paymentMethod => paymentMethod.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE);
         combinedPaymentMethods = _.map(combinedPaymentMethods, paymentMethod => ({
             ...paymentMethod,
             onPress: e => this.props.onPress(e, paymentMethod.accountType, paymentMethod.accountData, paymentMethod.isDefault, paymentMethod.methodID),
```

**Comment by bernhardoj** [comment]
@abdulrahuman5196 we need to filter the pending action only if we are online. That issue is similar to [this](https://github.com/Expensify/App/issues/13615). It would be a good addition with the filter (maybe can be handled on that issue too), but even though you filter it, the data is still there even after the request is completed until you relog in.

**Comment by abdulrahuman5196** [comment]
> we need to filter the pending action only if we are online

yes. This solves this issue. We need to have online check extra and will update proposal.
And as noted before UI filter would also solves this issue and other UI delay issues in this place due to network queues even when we are online.

> but even though you filter it, the data is still there even after the request is completed until you relog in.

This should be the stale data of payment method which happens at pusher failure(not sure on how likely this happens). But I agree in removing the stale data. This will also prevent any inconsistencies which can arise due to stale data.


**IMO** I think its **best implementing both the solutions** would solve this issue along with possible inconsistencies in displaying the Add payment message being displayed. 
But I would leave it to the reviewer to decide on having both solutions implemented or a single solution.

**Comment by jasperhuangg** [comment]
I just tried @bernhardoj's solution and it doesn't seem to work. If the successData set by the API response isn't updated, then I don't imagine setting it again in the BankAccounts action makes a difference.

@abdulrahuman5196 please update your proposal to include an offline check, and I'd be happy to re-review!

**Comment by bernhardoj** [comment]
Currently, the success data is received from the Pusher notification. By setting the success data to the API request, it will be set the next time we open the app (online).

So, here is my detailed explanation from my investigation based on the reproduction steps.

First, we perform a delete request of a payment. When we perform the request, the request will be saved to ONYXKEYS.PERSISTED_REQUESTS and does not get executed while offline.

https://github.com/Expensify/App/blob/c3923bd42abb066cc5d1a1fc3135eedb50030a54/src/libs/Network/SequentialQueue.js#L89-L105

Next, we reopen the app with online condition. Every time we open the app, there is always an API request of `SendPerformanceTiming` and attempt to subscribe to the Pusher. 

https://github.com/Expensify/App/blob/c3923bd42abb066cc5d1a1fc3135eedb50030a54/src/libs/Navigation/AppNavigator/AuthScreens.js#L95-L105

As you can see above, we subscribe first and then call the timing request. However, the subscription runs asynchronously, right? So, what would happen is the timing request completed before the subscription. Why the timing request matters? Because every time we perform an API request, we get all request from ONYXKEYS.PERSISTED_REQUESTS which also contains the delete payment request. When the delete payment request is completed, we wouldn't receive any success data from the Pusher as it is currently in the process of subscribing.

Pusher subscribe - ongoing
Delete payment - ongoing
Delete payment - success
Pusher subscribe - success

So, basically a race condition happens.

Regarding my solution, I just want to make sure you set the success data to `DeletePaymentBankAccount`, right? I'm afraid that you set the success data to the `DeleteWorkspace`.

Anyway, to test my hypothesis above, you can try to hold any API request waiting for the Pusher subscription completed. This is how I do it.
```diff
diff --git a/src/libs/Network/SequentialQueue.js b/src/libs/Network/SequentialQueue.js
index f37eb8422..7a3a24145 100644
--- a/src/libs/Network/SequentialQueue.js
+++ b/src/libs/Network/SequentialQueue.js
@@ -5,6 +5,7 @@ import * as NetworkStore from './NetworkStore';
 import ONYXKEYS from '../../ONYXKEYS';
 import * as ActiveClientManager from '../ActiveClientManager';
 import * as Request from '../Request';
+import PusherUtils from '../PusherUtils';
 
 let resolveIsReadyPromise;
 let isReadyPromise = new Promise((resolve) => {
@@ -25,12 +26,15 @@ let currentRequest = null;
  */
 function process() {
     const persistedRequests = PersistedRequests.getAll();
 
     // This sanity check is also a recursion exit point
-    if (NetworkStore.isOffline() || _.isEmpty(persistedRequests)) {
+    if (NetworkStore.isOffline() || _.isEmpty(persistedRequests) || !PusherUtils.isSubscribedToPrivateUserChannel()) {
         return Promise.resolve();
     }
 
     const task = _.reduce(persistedRequests, (previousRequest, request) => previousRequest.then(() => {
         currentRequest = Request.processWithMiddleware(request, true);
         return currentRequest;
diff --git a/src/libs/PusherUtils.js b/src/libs/PusherUtils.js
index 2cd614c60..5bb6fd139 100644
--- a/src/libs/PusherUtils.js
+++ b/src/libs/PusherUtils.js
@@ -3,7 +3,9 @@ import Log from './Log';
 import NetworkConnection from './NetworkConnection';
 import * as Pusher from './Pusher/pusher';
 import CONST from '../CONST';
+import * as SequentialQueue from './Network/SequentialQueue'
 
+let isSubscribed = false;
 /**
  * Abstraction around subscribing to private user channel events. Handles all logs and errors automatically.
  *
@@ -41,8 +43,18 @@ function subscribeToPrivateUserChannelEvent(eventName, accountID, onEvent) {
     }
     Pusher.subscribe(pusherChannelName, eventName, onEventPush, onPusherResubscribeToPrivateUserChannel)
         .catch(onSubscriptionFailed);
+    setTimeout(() => {
+        isSubscribed = true;
+        SequentialQueue.flush();
+    }, 5000);
+}
+
+function isSubscribedToPrivateUserChannel() {
+    return isSubscribed;
 }
 
 export default {
     subscribeToPrivateUserChannelEvent,
+    isSubscribedToPrivateUserChannel,
 };
```

And because we set a delay, it is very possible when you open the payment list, it will still show a blank list because the delete request has not complete yet, but after a while it will change to an empty list with a message, compared to the current condition where it is always showing a blank list until we relog in (because the onyx still contains the bank account with pending action of delete even the request is completed). That's the reason I say on my previous comment it will be a good addition to filter the list similar to the issue I also mentioned with the same root cause.

**Comment by bernhardoj** [comment]
No no, that's not a proposal by blocking the request. That's just for testing the race condition hypothesis.

> 2) Send API success data from server as part of API response also
Currently API's relaying on pusher events doesn't get any response data on the API response. We could start sending response data for API's if possible as a extra measure for pusher events failure.

That's basically what `successData` is for.

**Comment by thesahindia** [comment]
> In PaymentMethodList.js we have set the add payment method message to be displayed as 'ListEmptyComponent' which will only display when we provide empty data to FlatList. But in this case when we delete the payment method and go offline, we still have the payment method data with "pendingAction": "delete" value as noted below which will only go away after we get response from server(also request still in network queue as well). So we still pass a data with payment entry to FlatList and flatList doesn't show the add payment menthod message.

I have looked into it and the root cause is correct. Also checked the proposals (@jasperhuangg, just a heads up @bernhardoj's proposal also works) I need to check some test cases first, will comment here after that (maybe in the morning or on monday)

**Comment by thesahindia** [comment]
@abdulrahuman5196's proposal has an issue, the bank account shows up again if the user goes offline. I like @bernhardoj's [proposal](https://github.com/Expensify/App/issues/14046#issuecomment-1373566214) since it will clear the onyx data.

I think we should create a new issue for pusher to investigate more

C+ reviewed 🎀👀🎀

**Comment by abdulrahuman5196** [comment]
>@abdulrahuman5196's proposal has an issue, the bank account shows up again if the user goes offline.

@thesahindia  Yes. This is an existing issue even without my change, I mentioned this in previous comments as well. 
That's why i mentioned we would be required to add both the solutions. If we don't add UI filter, we would end up not showing the 'Add a payment method' message until when we get a response from 'DeletePaymentMethod' API(after response only Onyx will set successData) similar to this issue - https://github.com/Expensify/App/issues/13615

**Comment by jasperhuangg** [comment]
Hey guys,

I've read through all the proposals.

Thank you @bernhardoj for clearly explaining the issue at hand, and for providing a way to reliably test this out. After trying everything again I can confirm that it works, not sure why it didn't work before–I was definitely setting the successData in `DeletePaymentBankAccount`.

Your explanation also makes sense, I'm wondering if there are other API methods that this occurs for that we could fix in one fell swoop.

Thank you for your proposal as well @abdulrahuman5196. It also makes sense to me, but I'm going to have to go with @bernhardoj's solution since it better addresses the issue at its root.

Thanks again for all the discussion guys!

**Comment by bernhardoj** [comment]
Going to open the PR in a few hours. By the way, should I apply the filter similar to [this](https://github.com/Expensify/App/pull/14156/files) (so it will prevent future issue)?

cc: @jasperhuangg

**Comment by bernhardoj** [comment]
PR is up, but I am unable to test on iOS mWeb because of the sign in problem. Waiting to be whitelisted, but looks like John is taking a break until Thursday?

**Comment by jasperhuangg** [comment]
Hey @abdulrahuman5196, I made a mistake and chose the wrong proposal–I'm assigning you to the issue now, please open a PR and let me know when it's ready for review!

**Comment by bernhardoj** [comment]
Sorry, the bot adds the Reviewing label again because I updated the closed PR. I update the video and add an explanation why it does not work for @jasperhuangg.

**Comment by jasperhuangg** [comment]
Quick update: Just merged the PR :)

**Comment by muttmuure** [comment]
@jasperhuangg is this ready to be paid out or are we waiting on anything?

**Comment by muttmuure** [comment]
Clarifying what the regression test should be because it's not clear to me

**Comment by muttmuure** [comment]
Still chatting with Jasper, went overdue over the weekend

**Comment by muttmuure** [comment]
@bernhardoj @thesahindia can you apply for the job here? https://www.upwork.com/jobs/~014e025b9cf1e4c7a8

**Comment by muttmuure** [comment]
Great!

**Comment by muttmuure** [comment]
requested regression buddy check [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1674586795705739)

**Comment by bernhardoj** [comment]
@muttmuure hi, are we not eligible for the bonus? There is a small misunderstanding that delay the review 😄.

By the way, thanks for the payment.

cc: @jasperhuangg

**Comment by muttmuure** [comment]
Yep, looks good - adding a bonus for C and C+

**Comment by muttmuure** [comment]
Bonuses granted.
